### ISSUE TITLE
TypeError: crypto.randomBytes is not a function

### ISSUE THREAD

**Issue by oyinda-subair** [optimizely/javascript-sdk#150]
When activating an experiment as follows: 
`optimizelyClient.activate('experiment', 'user123')` returns `null` and  the logger shows this error `TypeError: crypto.randomBytes is not a function`

Code: 
```
optimizely.createInstance({ datafile: JSON.stringify(datafile),
  errorHandler: defaultErrorHandler,
  logger: defaultLogger.createLogger({
    logLevel: LOG_LEVEL.DEBUG
  }),
 })
```

```
const variation = optimizelyClient.activate('experiment', 'user123')
```
Further troubleshooting deduced that
```
optimizelyClient._sendImpressionEvent('experiment', variationKey, 'user123', {})
```
returns the error
`TypeError: crypto.randomBytes is not a function`

**Comment by oyinda-subair** [comment]
I had to resolve to using `import optimizely from '@optimizely/optimizely-sdk/lib/index.node'` instead of `import optimizely from '@optimizely/optimizely-sdk'` so that I can get through with what I needed to do.
Does anyone have an idea on what might be the issue?

**Comment by mikeng13** [comment]
Thanks for submitting this! 

Can you please let us know what version of Node you are using, as well as anything about your current setup/env that could help us more accurately reproduce this issue on our end?

Mike

**Comment by oyinda-subair** [comment]
`node -v` - `v8.10.0`
`framework` - Typescript
`NODE_ENV` - `development`

Is this information enough? @mikeng13

**Comment by mikeng13** [comment]
Yes, thanks! We'll look into this.

**Comment by spencerwilson-optimizely** [comment]
Hi @oyinda-subair !

Something strange seems to be happening for you... What version of `@optimizely/optimizely-sdk` are you using? Since we _just_ released 2.1.3 yesterday, you must be using 2.1.2 or earlier, which means that if you run your code with Node.js' `--inspect` and sprinkle some `debugger` statements around, you _should_ be able to observe this line in node_modules/\@optimizely/optimizely-sdk/dist/optimizely.node.js being reached:
```sh
$ ag 'require.*crypto' node_modules/\@optimizely/optimizely-sdk/dist/optimizely.node.js
10662:module.exports = require("crypto");
```
which shouldn't cause problems, and return a reference to Node.js' [crypto](https://nodejs.org/api/crypto.html#crypto_crypto) module, as long as it's evaluated in a Node.js environment.

One issue I see is that given you're working with TypeScript, since `@optimizely/optimizely-sdk` is a CommonJS module, it should be imported with TypeScript's `import = require()` syntax (described on [this page](https://www.typescriptlang.org/docs/handbook/modules.html)):
```
import Optimizely = require('@optimizely/optimizely-sdk');
```

That should ensure that `tsc` resolves the import to the correct module, which is the aforementioned CommonJS module, node_modules/\@optimizely/optimizely-sdk/dist/optimizely.node.js.

Additionally, we have a TypeScript project at Optimizely that doesn't seem to have this issue. For reference, here's its tsconfig:
```js
{
  "compilerOptions": {
    /* Basic Options */
    "target": "ES2017",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
    // "lib": [],                             /* Specify library files to be included in the compilation. */
    // "allowJs": true,                       /* Allow javascript files to be compiled. */
    // "checkJs": true,                       /* Report errors in .js files. */
    // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
    // "sourceMap": true,                     /* Generates corresponding '.map' file. */
    // "outFile": "./",                       /* Concatenate and emit output to single file. */
    "outDir": "dist",                        /* Redirect output structure to the directory. */
    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    // "composite": true,                     /* Enable project compilation */
    // "removeComments": true,                /* Do not emit comments to output. */
    // "noEmit": true,                        /* Do not emit outputs. */
    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    "strict": true,                           /* Enable all strict type-checking options. */
    // "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,              /* Enable strict null checks. */
    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                /* Report errors on unused locals. */
    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */

    /* Module Resolution Options */
    // "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                       /* List of folders to include type definitions from. */
    // "types": [],                           /* Type declaration files to be included in compilation. */
    // "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
    "esModuleInterop": true                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
    // "preserveSymlinks": true,              /* Do not resolve the real path of symlinks. */

    /* Source Map Options */
    // "sourceRoot": "./",                    /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "./",                       /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
  }
}
```
It has hardly any changes from the default that `tsc --init` produces.

If you don't figure it out, are you able to put up a minimal repository on GitHub that reproduces the problem? Thanks!

Cheers,
Spencer

**Comment by spencerwilson-optimizely** [comment]
@oyinda-subair , closing this due to inactivity, but don't hesitate to reopen if you observe an issue again.
